sdk/vppagent/dataplane/state/chain.go

```diff
@@ -0,0 +1,93 @@
+package state
+
+import (
+	"context"
+	"sync"
+
+	"github.com/golang/protobuf/ptypes/empty"
+
+	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/crossconnect"
+	"github.com/networkservicemesh/networkservicemesh/dataplane/pkg/apis/dataplane"
+)
+
+type chainKey string
+
+const (
+	chainState chainKey = "chainState"
+	current    chainKey = "current"
+)
+
+type chain struct {
+	source sync.Map
+}
+
+func (c *chain) SetCurrent(server dataplane.DataplaneServer) {
+	c.source.Store(current, server)
+}
+
+func (c *chain) Current() dataplane.DataplaneServer {
+	if v, ok := c.source.Load(current); ok {
+		if casted, ok := v.(dataplane.DataplaneServer); ok {
+			return casted
+		}
+	}
+	return nil
+}
+
+func (c *chain) Next(current dataplane.DataplaneServer) dataplane.DataplaneServer {
+	if stored, ok := c.source.Load(current); ok {
+		if v, ok := stored.(dataplane.DataplaneServer); ok {
+			return v
+		}
+	}
+	return nil
+}
+
+func WithChain(ctx context.Context, first dataplane.DataplaneServer, next []dataplane.DataplaneServer) context.Context {
```

Comment:
Why put the chain state here instead of as was done in the SDK with it living with the 'Chain' struct:
https://github.com/networkservicemesh/networkservicemesh/blob/master/sdk/endpoint/composite.go#L61

Please also include the Tracing/Logging features seen there as well.

Reply:
A big part of the point of this exercise is to make things look very much the same across the code base... so if you've developed an NSE, and start working on a Dataplane, things look familiar :)

Reply:
thanks, fixed